Recommend random IV generation for each encryption

Joey3000 opened the pull request:
Refs https://github.com/phpseclib/phpseclib/issues/907#issuecomment-168575823.

This is based on the following from [Wikipedia](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Initialization_vector_.28IV.29) (emphasis mine):

> An initialization vector has different security requirements than a key, so the IV usually does not need to be secret. **However, in most cases, it is important that an initialization vector is never reused under the same key.** For CBC and CFB, reusing an IV leaks some information about the first block of plaintext, and about any common prefix shared by the two messages. For OFB and CTR, reusing an IV completely destroys security.[6] This can be seen because both modes effectively create a bitstream that is XORed with the plaintext, and this bitstream is dependent on the password and IV only. Reusing a bitstream destroys security. In CBC mode, the IV must, in addition, be unpredictable at encryption time; in particular, the (previously) common practice of re-using the last ciphertext block of a message as the IV for the next message is insecure (for example, this method was used by SSL 2.0). If an attacker knows the IV (or the previous block of ciphertext) before he specifies the next plaintext, he can check his guess about plaintext of some block that was encrypted with the same key before (this is known as the TLS CBC IV attack).

Note: Since the documentation still refers to the 1.0 branch, the "[IV] defaults to all-NULLs if not explicitely defined" was not removed (see https://github.com/phpseclib/phpseclib/pull/926). The goal of this PR is to recommend defining the IV nevertheless.

terrafrost commented:
Looks good! I'll try to do it tonight or tomorrow as time permits.

Also, I appreciate the thoroughness of your code reviews - thanks! :)

Joey3000 commented:
No problem. Thank you for your great work on this library!

Joey3000 commented:
I've also added a recommended IV length, based on the infinite collection of wisdom that is [Wikipedia](https://en.wikipedia.org/wiki/Initialization_vector) (emphasis mine):
> The size of the IV is dependent on the cryptographic primitive used; for block ciphers, it is generally the cipher's block size. **Ideally, for encryption schemes, the unpredictable part of the IV has the same size as the key to compensate time/memory/data tradeoff attacks.**

Please let me know if what I added could be improved.

terrafrost commented:
I think that line in wikipedia is a little misleading. In the block cipher modes currently implemented by phpseclib the IV needs to be the block size. You can null pad it if it's too short and truncate it if it's too long but, at the end of the day, it needs to be the block size:

https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Common_modes

Maybe it's different in AES-GCM but phpseclib doesn't currently implement AES-GCM.

For stream ciphers it's different. Consider WEP. It used to have a 40-bit key and a 24-bit "IV" (that was really just concatenated to the key). ie. the IV is not the same size as the key. Later WEP was upgraded to support 104-bit keys but still used 24-bit "IV"'s (so total key length is 128-bits). Because the "IV"'s were so small the probability that keys would be re-used was greatly increased. So for that having an "IV" the same size as the key makes sense. But for block cipher modes it doesn't.

So I've nixed that thing. I've also replaced the comment with a code sample. Now, Crypt/Random.php is included for every mode save for ECB and `crypt_random_string($cipher->getBlockLength() >> 3)` is called. The advantage of that is that it may result in less people using an insecure RNG.

Joey3000 commented:
Great, your update (https://github.com/phpseclib/docs/commit/bc22d9bac1831174efc7e05837ce441717548709) is much better, tanks for feedback!

I'm closing this as fixed.
